blog/_posts/2019-09-06-profilers.md

```diff
@@ -0,0 +1,405 @@
+  ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎ 1   int.jl:53; +
+  ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎    1   int.jl:53; +
+  ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎    ╎  2   int.jl:53; +
+```
+
+We also see this has issues with identifying the stack in many cases (almost 40% failed!), similar to `perf`. One way to improve that (~10% failure rate) is to disable the omit-frame-pointer option, like we did before. Currently this requires this source code change:
+
+```diff
+diff --git a/src/jitlayers.cpp b/src/jitlayers.cpp
+index cf339d8a4b..f6e9c08e2c 100644
+--- a/src/jitlayers.cpp
++++ b/src/jitlayers.cpp
+@@ -3,9 +3,7 @@
+ #include "llvm-version.h"
+ #include "platform.h"
+ #include "options.h"
+-#if defined(_OS_WINDOWS_) || defined(_OS_FREEBSD_)
+ #  define JL_DISABLE_FPO
+-#endif
+ 
+ #include <iostream>
+ #include <sstream>
+```
+
+But otherwise, this seems fairly accurate, but yikes, what an arrow! I needed over 40 frames this way, and that makes this pretty hard to interpret.
+Spiky frame-graphs like this can be quite misleading to look at since the visual size (height) isn't relevant to performance.
+And it make it really hard to see where the time is being spent in aggregate.
+
+Julia implements a `noisefloor` option as well as `mincount` and `maxdepth` to help with some situations, but they do nothing for this example.
+The `mincount` option (also present in `Instruments.app` for min and max # samples; and in `perf` as a min percent threshold) drops any line that doesn't meet that threshold. The `maxdepth` option (aka `print_limit` in `perf`) cuts off the tree after some recursion limit.
+The `noisefloor` option is similar to `mincount`, but operates as a relative fraction of the parent frame instead of a constant threshold (I'm not aware of a similar option in other tools). This can help hide frames from the tree that contribute little to the immediate performance. This is based on the value engineering principle of focusing on the largest contributors first. For many problems, these can be greatly beneficial.
+
+However, none of these option are helpful for our tree above. We need more tools in our toolbox!
+
+Let's now examine how we can try to consolidate the recursive frames somehow. One option is the flat reports we looked at earlier.
+
+Here's that report with the counts corrected, and some extra information also now compiled for it:
+
+```
+julia> Profile.print(format=:flat, sortedby=:counts)
+ Count  Overhead File      Line Function
+   228         2 ./boot.jl  330 eval
+     2         2 ./int.jl    52 -
+    25        25 ./int.jl    53 +
+   321       155 ./REPL[2]    1 fib_r
+   514       426 ./REPL[1]    2 fib
+Total snapshots: 610
+```
+
+That's better! But now we've had to give up on the tree view, losing the ability to see the context of each call. Some of those functions are very common, like `+`, and it might not be sufficient to see them in aggregate like this. Can we combine the benefits of both views somehow?
```

Comment:
Since we don't know what changes you made to the code, it's unclear why `format=:tree` is now broken. (Hopefully it isn't?!?)